### PR TITLE
Skip-reranking when current_first/current_last is nil

### PR DESCRIPTION
**Pull request by bf4**
A take on 'first do no harm' when seeing the occasional

```
NoMethodError: undefined method `rank' for nil:NilClass (Most recent call first)


File /app/vendor/bundle/ruby/2.6.0/gems/ranked-model-0.4.4/lib/ranked-model/ranker.rb line 188 in rearrange_ranks
```

There are various possible strategies for handling a missing 'current_x'
- reset cache and retry
- make a best effort with what we know
- calculate 'freshness' of the current instance and do something when not fresh
  - have a pluggable strategy when no longer fresh
- try to lock the re-reranking

Related
- https://github.com/mixonic/ranked-model/issues/124
- https://github.com/mixonic/ranked-model/pull/117

Work done:
- just the failing test #149
- two commits: 1. failing test, 2. make it pass
  - strategy 1: skip cannot rearrange (this PR).
  - strategy 2: abort save #151
- unrelated misc  #150

### CODE REVIEW

**Comment by brendon** [comment]
My preference would be locking, perhaps with an advisory lock if you wanted to investigate that. Does this problem still occur with the latest version of ranked-model?

**Comment by brendon** [comment]
Sorry, obviously it does since that's mentioned in your error :)

How do we end up with no `current_first`? It would imply that `finder.first` returns nil which would be an empty list?

**Comment by bf4** [comment]
@brendon Here's our ranked model

```ruby
class JobProductionPlanSegment < ApplicationRecord
  include RankedModel
  ranks :sequence, with_same: :job_production_plan_id
```
The migration for it specified
```ruby
 t.integer :sequence, null: false
```

The database is pg 11

Here are some logs from a recent occurrence

all requests below are POST to /v1/job-production-plan-segments with the same job_production_plan_id


timestamp | server | pid | thread | post response status | post body sequence-position | notes
-------|----------|-----------|--------| ------|----------|---------
2019-07-21T13:37:26.112Z | web.1 | 451 | 70057487585260 | 201 | 0 | 
2019-07-21T13:37:26.149Z | web.1 | 836 | 70057503954640 | 500 | 1 |  failed
2019-07-21T13:37:26.234Z | web.1 | 451 | 70057487580040 | 201 | 3 | 
2019-07-21T13:37:26.352Z | web.1 | 836 | 70057504046960 | 201 | 2 | 
2019-07-21T13:37:26.456Z | web.1 | 451 | 70057487574620 | 201 | 4 | 
2019-07-21T13:37:35.848Z | web.1 | 451  | 70057487561080 | 201 | 1 | appears to be a successful retry of the failed request

so maybe sequence-position 0 hadn't been persisted yet?

File: lib/ranked-model/ranker.rb [hunk 1]
```diff
@@ -208,6 +214,11 @@ def rearrange_ranks
         end
       end
 
+      def handle_cannot_rearrange_ranks(message)
+        logger.warn "[RANKED_MODEL] #{message}. Skipping re-ranking of #{instance.model_name.name} #{instance.id}"
+        nil
```
Comment (bf4) [inline lib/ranked-model/ranker.rb]: returning nil as implemented will skip-reranking

alternatively, could have a strategy

```ruby
fail RankedModel::Unrankable, message, instance
```

Or, since we're in a `before_save` we can  'throw :abort' in the later rails versions, like

```ruby
  errors.add ranker.column, "Could not re-rank columns: #{message}"
  throw :abort
```

**Comment by bf4** [comment]
@brendon on reflection I decide it made sense to make failing the 'before-save' hook to be a first class thing, and then used that as the default strategy for handling current-first or current-last being missing.

If you want to point me to where the appropriate test would go and give a few pointers, I'd be happy to add it.

File: lib/ranked-model.rb [hunk 2]
```diff
@@ -29,11 +29,33 @@ def self.included base
...
+
+  def handle_ranking_failure(ranker, ranking_failed_reason)
+    return if ranking_failed_reason.nil?
+    errors.add ranker.column, ranking_failed_reason
+    @ranked_model_ranking_succeeded = false
+    abort_handle_ranking
+  end
+
+  if defined?(Rails) && Rails.version >= '5'
+    def abort_handle_ranking
```
Comment (bf4) [inline lib/ranked-model.rb]: is a reasonable way to discover the behavior. I think this is fine even if `ActiveSupport.halt_callback_chains_on_return_false` is true

File: lib/ranked-model.rb [hunk 3]
```diff
@@ -69,7 +95,7 @@ def ranks *args
     end
 
     def column_default ranker
-      column_defaults[ranker.name.to_s] if ActiveRecord::Base.connected?
+      column_defaults[ranker.name.to_s] if ranker_connection_class.connected?
```
Comment (bf4) [inline lib/ranked-model.rb]: not really part of this pr. is it's own commit to reduce ar::base envy

File: lib/ranked-model/ranker.rb [hunk 4]
```diff
@@ -208,6 +214,10 @@ def rearrange_ranks
         end
       end
 
+      def handle_cannot_rearrange_ranks(message)
+        throw :ranking_failed, "Could not re-rank: #{message}."
```
Comment (bf4) [inline lib/ranked-model/ranker.rb]: can be overridden to return nil to skip the re-ranking and not abort the save

default is to abort the save and add an error message

**Comment by bf4** [comment]
OMG, I wrote a failing test and made it pass.  But I don't think my code style is right for the repo.

File: lib/ranked-model.rb [hunk 5]
```diff
@@ -24,14 +25,49 @@ def self.included base
       }
     end
 
+    rails_version = ActiveRecord.version.to_s
+    if rails_version < '5.0'
+      base.module_eval <<-EOS.strip_heredoc, __FILE__, __LINE__ + 1
+        def handle_ranking_return_value(return_value)
+          @ranked_model_ranking_succeeded
```
Comment (bf4) [inline lib/ranked-model.rb]: some hoops that let us handle aborting the save, here no matter where in the ranker the problem originated, for the supported Rails versions

File: spec/duck-model/duck_spec.rb [hunk 6]
```diff
@@ -23,6 +23,38 @@
...
+        Duck.expects(:rankers).once.returns([age_ranker])
+        patched_duck_ranker = age_ranker.with(patched_duck)
+        age_ranker.expects(:with).once.returns(patched_duck_ranker)
+        class << patched_duck_ranker
+          def current_first
+            nil
+          end
+
+          def current_at_rank(_rank)
+            true # something truthy
```
Comment (bf4) [inline spec/duck-model/duck_spec.rb]: This is otherwise nil. It's need to be truthy for `rearrange_ranks` to be called

File: spec/duck-model/duck_spec.rb [hunk 7]
```diff
@@ -23,6 +23,32 @@
...
+          :name => 'Quacky',
+          :pond => 'Shin',
+          :age_position => 1,
+        )
+        age_ranker = Duck.rankers.detect {|ranker| ranker.name == :age }
+        Duck.expects(:rankers).once.returns([age_ranker])
+        patched_duck_ranker = age_ranker.with(patched_duck)
+        age_ranker.expects(:with).once.returns(patched_duck_ranker)
+        patched_duck_ranker.expects(:current_first).returns(nil)
+        patched_duck_ranker.expects(:current_at_rank).returns(:something_truthy)
```
Comment (bf4) [inline spec/duck-model/duck_spec.rb]: This is otherwise nil. It's need to be truthy for `rearrange_ranks` to be called

**Comment by brendon** [comment]
Hi @bf4, this is going to take me a while to look through. At first glance it seems like a mildly ungraceful solution to the problem that seems to deal with the symptoms rather than trying to figure out the cause. Nonetheless I'll star this to look at once I've cleared my other backlog of work :D

**Comment by bf4** [comment]
@brendon In that case, allow me to break this up into smaller prs

- strategy 1: skip cannot rearrange
- strategy 2: abort save
- unrelated misc

**Comment by bf4** [comment]
I replaced the commit in the PR with just the failing test

File: spec/duck-model/duck_spec.rb (hunk 7 above)
Comment (bf4) [reply spec/duck-model/duck_spec.rb]: return value never seems to need to be anything but truthy since it was introduced 8280511f674e7b9468168bf9bc5ba16537477fe5

File: lib/ranked-model/ranker.rb (hunk 1 above)
Comment (bf4) [reply lib/ranked-model/ranker.rb]: see #151

File: spec/duck-model/duck_spec.rb (hunk 7 above)
Comment (brendon) [reply spec/duck-model/duck_spec.rb]: That's true. Perhaps it was intended for future use. For it to return a boolean you'd probably want to implement a `current_at_rank?` method.

Comment (bf4) [reply spec/duck-model/duck_spec.rb]: @brendon So, the reason I'm looking at `current_at_rank` at all is because I wanted to reproduce the problem I experienced: creating a record, specifying a position, rearrange ranks failing due to current_first being nil

In order to do that, I discovered that in order for rearrange_ranks to be called in my situation, `current_at_rank` had to be truthy.  I looked at the method and determined, for mocking purposes, I didn't need to care about what it was doing or returning

I was a little curious about `current_at_rank` since I didn't see it used anywhere else or directly tested. So, I looked through the version history and found as noted above that it was introduced to replace the condition `(current_order.find do |rankable|  rankable.rank.nil? || rankable.rank == rank end)` which confirmed that its usage in `assure_unique_position` is only to return something truthy.

Interestingly, it's truthiness was replaced by `finder.except( :order ).where( ranker.column => rank ).first` where 'finder' is more or less equivalent to 'current_order'

now, why would `current_at_rank(rank)` be truthy when `current_first` is nil, when current_first is defined as `finder.first`. I'm not sure, but it could be a race condition or a query cache issue  or related to the `except` call (though I doubt it)

Comment (bf4) [reply spec/duck-model/duck_spec.rb]: I'm not sure exactly what the intended meaning of current_at_rank is, or if it changed during the refactor. I think it is that another record exists?

Comment (brendon) [reply spec/duck-model/duck_spec.rb]: Let's continue the discussion here: #149

**Comment by bf4** [comment]
Closing as stale
